src/Compilers/Core/Portable/Operations/BasicBlockBuilder.cs

```diff
@@ -0,0 +1,310 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.FlowAnalysis
+{
+    internal sealed class BasicBlockBuilder
```

Comment:
No changes in this file, it was just converted from nested type to a non-nested type so it can be used by newly added `CustomDataFlowAnalysis` type.

Reply:
>  so it can be used by newly added CustomDataFlowAnalysis type.

It concerns me that we need custom data flow anaysis for this feature.  Is there a doc/link to the new API this is exposing?

Reply:
> It concerns me that we need custom data flow anaysis for this feature

What is the concern? The whole idea behind the PR was to use control flow/dataflow analysis to identify category of functionality/performance/security issues which non-flow based analysis cannot identify. Additionally, I have executed this analyzer using AnalyzerRunner on Compilers.sln and confirmed the performance is identical to existing IDE analyzers.

> Is there a doc/link to the new API this is exposing?

There is currently no new API being exposed. It is based on an existing public API for ControlFlowGraph/BasicBlock and has interim internal APIs IDataFlowAnalyzer and CustomDataFlowAnalysis,which we hope to shape and release in future.

---
In reply to: [229794118](https://github.com/dotnet/roslyn/pull/30777#discussion_r229794118) [](ancestors = 229794118)

Reply:
The concern is that it's internal.  If this API is ok, then make it public and sit the IDE feature on top of hte public API.  If that takes time to ensure that the public API is sufficient, and in the meantime the feature is just in a feature-branch, then so be it.  There should not be any sort of rush that necessitates having to expose the guts of the compiler through linked-files insertion into the IDE.

Reply:
Its internal because it is just a shared internal utility (just like other utilities that we link to compiler and workspaces layer) that is used in two almost identical contexts in compiler layer and workspace layer, the only difference being its used for BasicBlockBuilder type in compiler's dataflow analysis and used for BasicBlock type in workspace layer's dataflow analysis. It does not use any other internal compiler APIs. If we decide that it is useful to expose such a dataflow walker utility as a public API, we will move to the public API.

> There should not be any sort of rush that necessitates having to expose the guts of the compiler through linked-files insertion into the IDE.  

Except that there are absolutely no internals being exposed here - it is just sharing identical code which is based on existing public APIs in the compiler layer.

---
In reply to: [229814693](https://github.com/dotnet/roslyn/pull/30777#discussion_r229814693) [](ancestors = 229814693)

Reply:
> Except that there are absolutely no internals being exposed here - it is just sharing identical code which is based on existing public APIs in the compiler layer.

How do customers use that code?  Why is the IDE blessed in terms of being able to do so?